libraries/tock-cells/src/map_cell.rs

```diff
@@ -17,23 +76,37 @@ pub struct MapCell<T> {
     // `.occupied` before calling `.val.get()` or `.val.assume_init()`. See
     // [mem::MaybeUninit](https://doc.rust-lang.org/core/mem/union.MaybeUninit.html).
     val: UnsafeCell<MaybeUninit<T>>,
-    occupied: Cell<bool>,
+    occupied: Cell<MapCellState>,
+}
+
+impl<T> Drop for MapCell<T> {
+    fn drop(&mut self) {
+        let state = self.occupied.get();
+        debug_assert!(state != InitBorrowed);
+        if state == Init {
+            unsafe {
+                // Safety: state being Init means that the MaybeUninit data was initted.
+                // the pointer to the data can never be used again as this
+                drop_in_place(self.val.get_mut().as_mut_ptr())
```

Comment:
This code isn't exercised in any tests; can you add a `MapCell<String>` test?

Reply:
It is exercised in test_drop. Just checked that removing this line does indeed cause that test to fail